### PR TITLE
Generative adversarial networks, adversarial training and examples

### PR DESCRIPTION
**Pull request by Andrewwango**
This PR includes the following:
- Losses: adversarial losses following DeblurGAN (Kupyn et al.), AmbientGAN (Bora et al.) and UAIR (Pajot et al.)
- Models: generator and discriminator models that can be used with the above losses: PatchGAN, ESRGAN, DCGAN
- Training: `AdversarialTrainer` for adversarial training, which is modular and builds off the new `deepinv.Trainer`
- Examples: demonstrating training various GANs for imaging inverse problems in the `deepinv` framework
- Tests for adversarial training

### Checks to be done before submitting your PR
- [x] `python3 -m pytest tests/` runs successfully.
- [x] `black .` runs successfully.
- [x] `make html` runs successfully (in the `docs/` directory).
- [x] Updated docstrings related to the changes (as applicable).
- [x] Added an entry to the [CHANGELOG.rst](../CHANGELOG.rst).

### CODE REVIEW

File: deepinv/training/adversarial.py [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+class AdversarialTrainer(Trainer):
+    """
+    Trainer class for training a reconstruction network using adversarial learning.
+
+    It overrides the :class:`deepinv.Trainer` class to provide the same functionality, whilst supporting training using adversarial losses. Note that the forward pass remains the same.
+
+    The usual reconstruction model corresponds to the generator model in an adversarial framework, which is trained using losses specified in the ``losses`` argument.
+    Additionally, a discriminator model ``D`` is also jointly trained using the losses provided in ``losses_d``. The adversarial losses themselves are defined in the ``deepinv.loss.adversarial`` module.
+
+    See ``deepinv.examples.adversarial_learning`` for usage.
```
Comment (samuro95) [inline deepinv/training/adversarial.py:83]: Could we have here a little snippet of code ?

File: deepinv/training/adversarial.py [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+                            else self.logs_losses_eval[k + len(self.losses)]
+                        )
+                        current_log.update(loss.detach().cpu().numpy())
+                        cur_loss = current_log.avg
+                        logs[l.__class__.__name__] = cur_loss
+
+            if train:
+                loss_total_d.backward()
+
+                self.clip_grad_D()
```
Comment (samuro95) [inline deepinv/training/adversarial.py:211]: Maybe we can also visualize discriminator norm in the same way that we do for the generator ?

**Review by samuro95 — APPROVED** [review]

**Review by samuro95 — CHANGES_REQUESTED** [review]
Can you remove the demo ipynb file ?

File: deepinv/training/adversarial.py [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+class AdversarialScheduler:
+    r"""Scheduler for adversarial training that encapsulates both generator and discriminator's schedulers.
+
+    :param LRScheduler scheduler_g: generator's torch scheduler
+    :param LRScheduler scheduler_d: discriminator's torch scheduler
+    """
+
+    def __init__(self, scheduler_g: LRScheduler, scheduler_d: LRScheduler):
+        self.G = scheduler_g
+        self.D = scheduler_d
```
Comment (matthieutrs) [inline deepinv/training/adversarial.py:69]:
```suggestion
        self.scheduler_g = scheduler_g
        self.scheduler_d = scheduler_d
```

File: deepinv/training/adversarial.py [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+
+    def get_last_lr(self):
+        return self.G.get_last_lr()
```
Comment (matthieutrs) [inline deepinv/training/adversarial.py:72]:
```suggestion
        return self.scheduler_g.get_last_lr()
```

File: deepinv/training/adversarial.py [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+
+    def step(self):
+        self.G.step()
+        self.D.step()
```
Comment (matthieutrs) [inline deepinv/training/adversarial.py:76]:
```suggestion
        self.scheduler_g.step()
        self.scheduler_d.step()
```

File: deepinv/training/adversarial.py [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+    """
+
+    def __init__(self, scheduler_g: LRScheduler, scheduler_d: LRScheduler):
+        self.G = scheduler_g
+        self.D = scheduler_d
+
+    def get_last_lr(self):
+        return self.G.get_last_lr()
+
+    def step(self):
```
Comment (matthieutrs) [inline deepinv/training/adversarial.py:74]:
```suggestion
    def step(self):
    r"""Performs a step on both generator and discriminator schedulers."""
```

File: deepinv/training/adversarial.py [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+
+    :param LRScheduler scheduler_g: generator's torch scheduler
+    :param LRScheduler scheduler_d: discriminator's torch scheduler
+    """
+
+    def __init__(self, scheduler_g: LRScheduler, scheduler_d: LRScheduler):
+        self.G = scheduler_g
+        self.D = scheduler_d
+
+    def get_last_lr(self):
```
Comment (matthieutrs) [inline deepinv/training/adversarial.py:71]:
```suggestion
    def get_last_lr(self):
    r"""Get last learning rates from the generator scheduler."""
```

**Review by matthieutrs — CHANGES_REQUESTED** [review]

**Comment by Andrewwango** [comment]
TODO: change example to use new `datasets.Urban100`

**Comment by tachella** [comment]
@Andrewwango, could you have a look at the corrections? Let me know if you need any help or have any doubts :)

**Review by tachella — CHANGES_REQUESTED** [review]

File: deepinv/loss/adversarial/base.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+from torch import Tensor
+from deepinv.loss.loss import Loss
+
+
+class DiscriminatorMetric:
+    r"""Generic GAN discriminator metric building block.
+
+    Compares discriminator output with labels depending on if the image should be real or not.
+
+    :param nn.Module metric: loss with which to compare outputs, defaults to nn.MSELoss()
```
[the remark this answers is not in the document]
Comment (Andrewwango) [reply deepinv/loss/adversarial/base.py:14]: This implementation follows LSGAN but DiscriminatorMetric can be overriden for whatever the user wants.

File: examples/adversarial-learning/demo_gan_imaging.py [hunk 9]
```diff
@@ -0,0 +1,346 @@
...
+
+download_and_extract_archive(
+    "https://huggingface.co/datasets/eugenesiow/Urban100/resolve/main/data/Urban100_HR.tar.gz?download=true",
+    "Urban100",
+    filename="Urban100_HR.tar.gz",
+    md5="65d9d84a34b72c6f7ca1e26a12df1e4c",
+)
+
+train_dataset, test_dataset = random_split(
+    ImageFolder(
```
[the remark this answers is not in the document]
Comment (Andrewwango) [reply examples/adversarial-learning/demo_gan_imaging.py:70]: I think this is orthogonal to the PR and should be picked up later

File: examples/adversarial-learning/demo_gan_imaging.py [hunk 10]
```diff
@@ -0,0 +1,346 @@
...
+loss_d = adversarial.SupAdversarialDiscriminatorLoss(device=device)
+
+
+# %%
+# As before, we can now train our models. Since inference is very
+# slow for CSGM/AmbientGAN as it requires an optimisation, we only do one
+# evaluation at the end. Note the train PSNR is meaningless as this
+# generative model is trained on random latents.
+#
+
```
[the remark this answers is not in the document]
Comment (Andrewwango) [reply examples/adversarial-learning/demo_gan_imaging.py:318]: This would be hard as AmbientGAN is quite tricky to train and I think it would struggle to get good results on this dummy example. I included a pretrained example with DeblurGAN earlier on in the script

**Review by tachella — APPROVED** [review]
